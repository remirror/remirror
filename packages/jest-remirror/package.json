{
  "name": "jest-remirror",
  "version": "3.0.0-beta.6",
  "description": "A text editor for react built with prosemirror",
  "keywords": [
    "prosemirror",
    "react",
    "text editor"
  ],
  "homepage": "https://github.com/remirror/remirror/tree/HEAD/packages/jest-remirror",
  "repository": {
    "type": "git",
    "url": "https://github.com/remirror/remirror.git",
    "directory": "packages/jest-remirror"
  },
  "license": "MIT",
  "contributors": [
    "Ifiok Jr. <ifiokotung@gmail.com>"
  ],
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist-types/index.d.ts",
      "import": "./dist/jest-remirror.js",
      "require": "./dist/jest-remirror.cjs"
    },
    "./environment": {
      "types": "./dist-types/environment.d.ts",
      "import": "./environment/dist/jest-remirror-environment.js",
      "require": "./environment/dist/jest-remirror-environment.cjs"
    },
    "./package.json": "./package.json"
  },
  "main": "./dist/jest-remirror.cjs",
  "module": "./dist/jest-remirror.js",
  "types": "./dist-types/index.d.ts",
  "files": [
    "dist",
    "dist-types",
    "environment"
  ],
  "scripts": {
    "build": "remirror-cli build"
  },
  "dependencies": {
    "@babel/runtime": "^7.22.3",
    "@remirror/core": "3.0.0-beta.6",
    "@remirror/dom": "3.0.0-beta.6",
    "@remirror/pm": "3.0.0-beta.4",
    "@remirror/preset-core": "3.0.0-beta.6",
    "@testing-library/dom": "^9.3.0",
    "@types/sanitize-html": "^2.9.0",
    "jest-prosemirror": "3.0.0-beta.4",
    "sanitize-html": "^2.10.0",
    "tiny-warning": "^1.0.3"
  },
  "devDependencies": {
    "@remirror/cli": "1.1.0-beta.0",
    "@types/node": "^16.18.34",
    "jest": "^29.5.0"
  },
  "peerDependencies": {
    "@types/node": "*",
    "jest": ">= 24.9.0"
  },
  "peerDependenciesMeta": {
    "@types/node": {
      "optional": true
    },
    "jest": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public",
    "exports": {
      ".": {
        "types": "./dist/jest-remirror.d.ts",
        "import": "./dist/jest-remirror.js",
        "require": "./dist/jest-remirror.cjs"
      },
      "./environment": {
        "types": "./environment/dist/jest-remirror-environment.d.ts",
        "import": "./environment/dist/jest-remirror-environment.js",
        "require": "./environment/dist/jest-remirror-environment.cjs"
      },
      "./package.json": "./package.json"
    },
    "main": "./dist/jest-remirror.cjs",
    "module": "./dist/jest-remirror.js",
    "type": "module",
    "types": "./dist/jest-remirror.d.ts"
  },
  "@remirror": {
    "tsconfigs": {
      "src": {
        "compilerOptions": {
          "types": [
            "node",
            "jest"
          ]
        }
      }
    }
  },
  "preconstruct": {
    "entrypoints": [
      "index.ts",
      "environment.ts"
    ]
  }
}
