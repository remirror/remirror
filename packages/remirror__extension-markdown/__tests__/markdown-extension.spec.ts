import { extensionValidityTest, renderEditor } from 'jest-remirror';
import {
  BoldExtension,
  BulletListExtension,
  HeadingExtension,
  ItalicExtension,
  OrderedListExtension,
  TaskListExtension,
} from 'remirror/extensions';

import { MarkdownExtension } from '../';

extensionValidityTest(MarkdownExtension);

describe('commands.insertMarkdown', () => {
  it('can insert blocks', () => {
    const editor = renderEditor([
      new HeadingExtension(),
      new BoldExtension(),
      new ItalicExtension(),
      new MarkdownExtension(),
    ]);
    const { doc, p } = editor.nodes;
    const { bold, italic } = editor.marks;
    const { heading } = editor.attributeNodes;
    const h1 = heading({ level: 1 });

    editor.add(doc(p('Content<cursor>')));

    editor.chain.insertMarkdown('# This is a heading').selectText('end').run();
    expect(editor.state.doc).toEqualProsemirrorNode(doc(p('Content'), h1('This is a heading')));

    editor.commands.insertMarkdown('A paragraph _with_ **formatting**');
    expect(editor.state.doc).toEqualProsemirrorNode(
      doc(
        p('Content'),
        h1('This is a heading'),
        p('A paragraph ', italic('with'), ' ', bold('formatting')),
      ),
    );
  });

  it('can replace blocks', () => {
    const editor = renderEditor([new HeadingExtension(), new MarkdownExtension()]);
    const { doc, p } = editor.nodes;
    const { heading } = editor.attributeNodes;
    const h1 = heading({ level: 1 });

    editor.add(doc(p('<cursor>')));

    editor.chain.insertMarkdown('# This is a heading').selectText('end').run();
    expect(editor.state.doc).toEqualProsemirrorNode(doc(h1('This is a heading')));
  });

  it('can insert multiple blocks', () => {
    const editor = renderEditor([new HeadingExtension(), new MarkdownExtension()]);
    const { doc, p } = editor.nodes;
    const { heading } = editor.attributeNodes;
    const h1 = heading({ level: 1 });

    editor.add(doc(p('<cursor>')));

    editor.chain.insertMarkdown('# Heading\nAnd content').run();
    expect(editor.state.doc).toEqualProsemirrorNode(doc(h1('Heading'), p('And content')));
  });

  it('can insert marks', () => {
    const editor = renderEditor([new BoldExtension(), new MarkdownExtension()]);
    const { doc, p } = editor.nodes;
    const { bold } = editor.marks;

    editor.add(doc(p('Content <cursor>')));

    editor.chain.insertMarkdown('**is bold.**').selectText('end').run();
    expect(editor.state.doc).toEqualProsemirrorNode(doc(p('Content ', bold('is bold.'))));
    expect(editor.dom).toMatchSnapshot();
  });

  it('can insert lists', () => {
    const editor = renderEditor([
      new BulletListExtension(),
      new OrderedListExtension(),
      new TaskListExtension(),
      new MarkdownExtension(),
    ]);
    const { doc, p, bulletList: ul, orderedList: ol, taskList, listItem: li } = editor.nodes;
    const { taskListItem } = editor.attributeNodes;

    editor.add(doc(p('<cursor>')));

    // Intentionally the same as the output from getMarkdown helper
    const tabSpace = `    `;
    const markdown = `Bullet list

*   Item 1
${tabSpace}
*   Item 2
${tabSpace}

Ordered list

1.  Item 1
${tabSpace}
2.  Item 2
${tabSpace}

Task list

- [x] Item 1

- [ ] Item 2`;

    editor.chain.insertMarkdown(markdown).selectText('end').run();
    const expected = doc(
      //
      p('Bullet list'),
      ul(
        //
        li(p('Item 1')),
        li(p('Item 2')),
      ),
      p('Ordered list'),
      ol(
        //
        li(p('Item 1')),
        li(p('Item 2')),
      ),
      p('Task list'),
      taskList(
        //
        taskListItem({ checked: true })(p('Item 1')),
        taskListItem({ checked: false })(p('Item 2')),
      ),
    );

    expect(editor.state.doc).toEqualProsemirrorNode(expected);
    expect(editor.dom).toMatchSnapshot();
  });

  it('does not replace marks', () => {
    const editor = renderEditor([new BoldExtension(), new MarkdownExtension()]);
    const { doc, p } = editor.nodes;
    const { bold } = editor.marks;

    editor.add(doc(p('<cursor>')));

    editor.chain.insertMarkdown('**is bold.**').selectText('end').run();
    expect(editor.state.doc).toEqualProsemirrorNode(doc(p(bold('is bold.'))));
    expect(editor.dom).toMatchSnapshot();
  });

  it('can insert marks as blocks', () => {
    const editor = renderEditor([new BoldExtension(), new MarkdownExtension()]);
    const { doc, p } = editor.nodes;
    const { bold } = editor.marks;

    editor.add(doc(p('Content <cursor>')));

    editor.chain
      .insertMarkdown('**is bold.**', { alwaysWrapInBlock: true })
      .selectText('end')
      .run();
    expect(editor.state.doc).toEqualProsemirrorNode(doc(p('Content '), p(bold('is bold.'))));
    expect(editor.dom).toMatchSnapshot();
  });
});

describe('helpers.getMarkdown', () => {
  it('returns the expected markdown content for lists', () => {
    const editor = renderEditor([
      new BulletListExtension(),
      new OrderedListExtension(),
      new TaskListExtension(),
      new MarkdownExtension(),
    ]);
    const { doc, p, bulletList: ul, orderedList: ol, taskList, listItem: li } = editor.nodes;
    const { taskListItem } = editor.attributeNodes;

    editor.add(
      doc(
        //
        p('Bullet list'),
        ul(
          //
          li(p('Item 1')),
          li(p('Item 2')),
        ),
        p('Ordered list'),
        ol(
          //
          li(p('Item 1')),
          li(p('Item 2')),
        ),
        p('Task list'),
        taskList(
          //
          taskListItem({ checked: true })(p('Item 1')),
          taskListItem({ checked: false })(p('Item 2')),
        ),
      ),
    );

    const tabSpace = `    `;
    const expectedMarkdown = `Bullet list

*   Item 1
${tabSpace}
*   Item 2
${tabSpace}

Ordered list

1.  Item 1
${tabSpace}
2.  Item 2
${tabSpace}

Task list

- [x] Item 1

- [ ] Item 2`;

    expect(editor.helpers.getMarkdown()).toEqual(expectedMarkdown);
  });
});
