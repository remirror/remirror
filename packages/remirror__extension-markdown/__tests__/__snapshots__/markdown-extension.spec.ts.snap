// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands.insertMarkdown can insert basic tables from markdown 1`] = `
<div
  aria-label=""
  aria-multiline="true"
  class="ProseMirror remirror-editor"
  contenteditable="true"
  role="textbox"
  translate="no"
>
  <div
    class="tableWrapper"
  >
    <table
      style="min-width: 50px;"
    >
      <colgroup>
        <col />
        <col />
      </colgroup>
      <tbody>
        <tr>
          <th>
            <p>
              Column 1
            </p>
          </th>
          <th>
            <p>
              Column 2
            </p>
          </th>
        </tr>
        <tr>
          <td>
            <p>
              Row 1, Column 1
            </p>
          </td>
          <td>
            <p>
              Row 1, Column 2
            </p>
          </td>
        </tr>
        <tr>
          <td>
            <p>
              Row 2, Column 1
            </p>
          </td>
          <td>
            <p>
              Row 2, Column 2
            </p>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`commands.insertMarkdown can insert lists 1`] = `
<div
  aria-label=""
  aria-multiline="true"
  class="ProseMirror remirror-editor"
  contenteditable="true"
  role="textbox"
  translate="no"
>
  <p>
    Bullet list
  </p>
  <ul>
    <li>
      <p>
        Item 1
      </p>
    </li>
    <li>
      <p>
        Item 2
      </p>
    </li>
  </ul>
  <p>
    Ordered list
  </p>
  <ol>
    <li>
      <p>
        Item 1
      </p>
    </li>
    <li>
      <p>
        Item 2
      </p>
    </li>
  </ol>
  <p>
    Task list
  </p>
  <ul
    data-task-list=""
  >
    <li
      class="remirror-list-item-with-custom-mark"
      data-checked=""
      data-task-list-item=""
    >
      <label
        class="remirror-list-item-marker-container"
      >
        <input
          class="remirror-list-item-checkbox"
          type="checkbox"
        />
      </label>
      <div>
        <p>
          Item 1
        </p>
      </div>
    </li>
    <li
      class="remirror-list-item-with-custom-mark"
      data-task-list-item=""
    >
      <label
        class="remirror-list-item-marker-container"
      >
        <input
          class="remirror-list-item-checkbox"
          type="checkbox"
        />
      </label>
      <div>
        <p>
          Item 2
        </p>
      </div>
    </li>
  </ul>
</div>
`;

exports[`commands.insertMarkdown can insert marks 1`] = `
<div
  aria-label=""
  aria-multiline="true"
  class="ProseMirror remirror-editor"
  contenteditable="true"
  role="textbox"
  translate="no"
>
  <p>
    Content 
    <strong>
      is bold.
    </strong>
  </p>
</div>
`;

exports[`commands.insertMarkdown can insert marks as blocks 1`] = `
<div
  aria-label=""
  aria-multiline="true"
  class="ProseMirror remirror-editor"
  contenteditable="true"
  role="textbox"
  translate="no"
>
  <p>
    Content 
  </p>
  <p>
    <strong>
      is bold.
    </strong>
  </p>
</div>
`;

exports[`commands.insertMarkdown can insert react tables from markdown 1`] = `
<div
  aria-label=""
  aria-multiline="true"
  class="ProseMirror remirror-editor"
  contenteditable="true"
  role="textbox"
  translate="no"
>
  <div
    class="remirror-table-show-controllers"
  >
    <table
      class="remirror-table remirror-table-with-controllers"
      data-controllers-injected=""
      style="min-width: 30px;"
    >
      <colgroup
        class="remirror-table-colgroup"
      >
        <col />
        <col />
        <col />
      </colgroup>
      <tbody
        class="remirror-table-tbody-with-controllers"
      >
        <tr>
          <th
            class="remirror-table-controller remirror-controllers-toggle"
            contenteditable="false"
            data-controller-cell=""
          >
            <div
              class="remirror-table-controller-wrapper"
              contenteditable="false"
            >
              <div
                contenteditable="false"
              />
              <div
                class="remirror-table-controller-trigger-area"
              />
              <div
                class="remirror-table-controller-trigger-area"
              />
              <div
                class="remirror-table-controller-mark-row-corner"
              />
              <div
                class="remirror-table-controller-mark-column-corner"
              />
            </div>
          </th>
          <th
            class="remirror-table-controller remirror-controllers-toggle"
            contenteditable="false"
            data-controller-cell=""
          >
            <div
              class="remirror-table-controller-wrapper"
              contenteditable="false"
            >
              <div
                contenteditable="false"
              />
              <div
                class="remirror-table-controller-trigger-area"
              />
              <div
                class="remirror-table-controller-trigger-area"
              />
              <div
                class="remirror-table-controller-mark-row-corner"
              />
              <div
                class="remirror-table-controller-mark-column-corner"
              />
            </div>
          </th>
          <th
            class="remirror-table-controller remirror-controllers-toggle"
            contenteditable="false"
            data-controller-cell=""
          >
            <div
              class="remirror-table-controller-wrapper"
              contenteditable="false"
            >
              <div
                contenteditable="false"
              />
              <div
                class="remirror-table-controller-trigger-area"
              />
              <div
                class="remirror-table-controller-trigger-area"
              />
              <div
                class="remirror-table-controller-mark-row-corner"
              />
              <div
                class="remirror-table-controller-mark-column-corner"
              />
            </div>
          </th>
        </tr>
        <tr>
          <th
            class="remirror-table-controller remirror-controllers-toggle"
            contenteditable="false"
            data-controller-cell=""
          >
            <div
              class="remirror-table-controller-wrapper"
              contenteditable="false"
            >
              <div
                contenteditable="false"
              />
              <div
                class="remirror-table-controller-trigger-area"
              />
              <div
                class="remirror-table-controller-trigger-area"
              />
              <div
                class="remirror-table-controller-mark-row-corner"
              />
              <div
                class="remirror-table-controller-mark-column-corner"
              />
            </div>
          </th>
          <th>
            <p>
              Column 1
            </p>
          </th>
          <th>
            <p>
              Column 2
            </p>
          </th>
        </tr>
        <tr>
          <th
            class="remirror-table-controller remirror-controllers-toggle"
            contenteditable="false"
            data-controller-cell=""
          >
            <div
              class="remirror-table-controller-wrapper"
              contenteditable="false"
            >
              <div
                contenteditable="false"
              />
              <div
                class="remirror-table-controller-trigger-area"
              />
              <div
                class="remirror-table-controller-trigger-area"
              />
              <div
                class="remirror-table-controller-mark-row-corner"
              />
              <div
                class="remirror-table-controller-mark-column-corner"
              />
            </div>
          </th>
          <td>
            <p>
              Row 1, Column 1
            </p>
          </td>
          <td>
            <p>
              Row 1, Column 2
            </p>
          </td>
        </tr>
        <tr>
          <th
            class="remirror-table-controller remirror-controllers-toggle"
            contenteditable="false"
            data-controller-cell=""
          >
            <div
              class="remirror-table-controller-wrapper"
              contenteditable="false"
            >
              <div
                contenteditable="false"
              />
              <div
                class="remirror-table-controller-trigger-area"
              />
              <div
                class="remirror-table-controller-trigger-area"
              />
              <div
                class="remirror-table-controller-mark-row-corner"
              />
              <div
                class="remirror-table-controller-mark-column-corner"
              />
            </div>
          </th>
          <td>
            <p>
              Row 2, Column 1
            </p>
          </td>
          <td>
            <p>
              Row 2, Column 2
            </p>
          </td>
        </tr>
      </tbody>
    </table>
    <div />
  </div>
</div>
`;

exports[`commands.insertMarkdown does not replace marks 1`] = `
<div
  aria-label=""
  aria-multiline="true"
  class="ProseMirror remirror-editor"
  contenteditable="true"
  role="textbox"
  translate="no"
>
  <p>
    <strong>
      is bold.
    </strong>
  </p>
</div>
`;
